overlord/snapstate/snapmgr.go

```diff
@@ -975,18 +1003,36 @@ func (m *SnapManager) undoLinkSnap(t *state.Task, _ *tomb.Tomb) error {
 	if err != nil {
 		return err
 	}
-	var hadCandidate bool
-	err = t.Get("had-candidate", &hadCandidate)
-	if err != nil && err != state.ErrNoState {
-		return err
+
+	isRevert := false
+	oldCandidateIndex := -1
+	if ss.HasABA() {
+		isRevert = ss.IsRevert()
+		err = t.Get("old-candidate-index", &oldCandidateIndex)
+		if err != nil {
+			return err
+		}
+	} else {
+		err = t.Get("had-candidate", &isRevert)
+		if err != nil && err != state.ErrNoState {
+			return err
+		}
 	}
 
 	// relinking of the old snap is done in the undo of unlink-current-snap
 	currentIndex := snapst.findIndex(snapst.Current)
 	if currentIndex < 0 {
 		return fmt.Errorf("internal error: cannot find revision %d in %v for undoing the added revision", ss.SideInfo.Revision, snapst.Sequence)
 	}
-	if !hadCandidate {
+	if ss.HasABA() {
+		if oldCandidateIndex < 0 {
+			snapst.Sequence = append(snapst.Sequence[:currentIndex], snapst.Sequence[currentIndex+1:]...)
```

Comment:
don't we have a clearer sense now that this was appended ? should we make it clearer here, or do we expect to do stranger things at some point?

some more comments around these blocks could help